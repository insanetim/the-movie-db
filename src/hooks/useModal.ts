import { nanoid } from '@reduxjs/toolkit'
import { useCallback } from 'react'
import { useModalsContext } from 'src/contexts/ModalsProvider'
import { Modal } from 'src/contexts/ModalsProvider/types'

const useModal = (initModalId?: string) => {
  const { dispatch } = useModalsContext()

  const modalId = initModalId ?? nanoid()

  const openModal = useCallback(
    (modal: Omit<Modal, 'modalId'>) => {
      dispatch({ payload: { ...modal, modalId }, type: 'OPEN_MODAL' })
    },
    [dispatch, modalId]
  )

  const closeModal = useCallback(() => {
    dispatch({ payload: modalId, type: 'CLOSE_MODAL' })
  }, [dispatch, modalId])

  return { closeModal, openModal }
}

export default useModal
