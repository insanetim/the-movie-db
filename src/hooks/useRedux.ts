import { TypedUseSelectorHook, useDispatch, useSelector } from 'react-redux'

import type { AppDispatch, RootState } from '../store'

type DispatchFunc = () => AppDispatch
const useAppDispatch: DispatchFunc = useDispatch
const useAppSelector: TypedUseSelectorHook<RootState> = useSelector

export { useAppDispatch, useAppSelector }
