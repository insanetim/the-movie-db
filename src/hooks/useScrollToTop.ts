import { useEffect } from 'react'
import { useLocation } from 'react-router-dom'

const useScrollToTop = () => {
  const { pathname } = useLocation()

  useEffect(() => {
    const timeout = setTimeout(() => {
      window.scrollTo(0, 0)
    }, 0)

    return () => {
      clearTimeout(timeout)
    }
  }, [pathname])
}

export default useScrollToTop
