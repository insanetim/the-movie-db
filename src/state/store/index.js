import { configureStore } from '@reduxjs/toolkit'
import { createLogicMiddleware } from 'redux-logic'

import httpClient from 'src/api/httpClient'
import rootReducer from './reducer'
import appOperations from '../app/operations'
import sessionOperations from '../session/operations'
import dashboardOperations from '../dashboard/operations'
import listsOperations from '../lists/operations'
import watchlistOperations from '../watchlist/operations'
import favoritesOperations from '../favorites/operations'
import movieOperations from '../movie/operations'

const operations = [
  ...appOperations,
  ...sessionOperations,
  ...dashboardOperations,
  ...listsOperations,
  ...watchlistOperations,
  ...favoritesOperations,
  ...movieOperations
]
const operationsDependencies = { httpClient }
const logicMiddleware = createLogicMiddleware(operations, operationsDependencies)

const store = configureStore({
  reducer: rootReducer,
  middleware: getDefaultMiddleware =>
    getDefaultMiddleware({
      serializableCheck: null
    }).concat(logicMiddleware),
  devTools: process.env.NODE_ENV !== 'production'
})

export default store
