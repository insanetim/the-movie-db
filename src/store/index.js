import { configureStore } from '@reduxjs/toolkit'
import { createLogicMiddleware } from 'redux-logic'

import httpClient from 'src/api/httpClient'
import appReducer from './app'
import sessionReducer from './session'
import dashboardReducer from './dashboard'
import listsReducer from './lists'
import watchlistReducer from './watchlist'
import favoritesReducer from './favorites'
import movieReducer from './movie'
import appOperations from './app/operations'
import sessionOperations from './session/operations'
import dashboardOperations from './dashboard/operations'
import listsOperations from './lists/operations'
import watchlistOperations from './watchlist/operations'
import favoritesOperations from './favorites/operations'
import movieOperations from './movie/operations'

const operations = [
  ...appOperations,
  ...sessionOperations,
  ...dashboardOperations,
  ...listsOperations,
  ...watchlistOperations,
  ...favoritesOperations,
  ...movieOperations
]
const operationsDependencies = { httpClient }
const logicMiddleware = createLogicMiddleware(operations, operationsDependencies)

const store = configureStore({
  reducer: {
    app: appReducer,
    session: sessionReducer,
    dashboard: dashboardReducer,
    lists: listsReducer,
    watchlist: watchlistReducer,
    favorites: favoritesReducer,
    movie: movieReducer
  },
  middleware: getDefaultMiddleware => getDefaultMiddleware().concat(logicMiddleware),
  devTools: process.env.NODE_ENV !== 'production'
})

export default store
