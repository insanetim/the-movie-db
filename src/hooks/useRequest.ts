import type { AsyncThunkAction } from '@reduxjs/toolkit'

import { type Dispatch, unwrapResult } from '@reduxjs/toolkit'
import { useCallback, useState } from 'react'

import { useAppDispatch } from './useRedux'

declare type AsyncThunkConfig = {
  dispatch?: Dispatch
  extra?: unknown
  fulfilledMeta?: unknown
  pendingMeta?: unknown
  rejectValue?: unknown
  rejectedMeta?: unknown
  serializedErrorType?: unknown
  state?: unknown
}

const useRequest = (initialLoading = true) => {
  const dispatch = useAppDispatch()
  const [loading, setLoading] = useState(initialLoading)
  const [error, setError] = useState<null | string>(null)

  const request = useCallback(
    async (action: AsyncThunkAction<unknown, unknown, AsyncThunkConfig>) => {
      setLoading(true)
      setError(null)

      try {
        unwrapResult(await dispatch(action))
      } catch (e) {
        setError(e as string)
      }

      setLoading(false)
    },
    [dispatch]
  )

  return { error, loading, request }
}

export default useRequest
