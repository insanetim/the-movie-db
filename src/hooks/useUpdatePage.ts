import { AsyncThunkAction, Dispatch } from '@reduxjs/toolkit'
import { SetURLSearchParams } from 'react-router-dom'
import getParams from 'src/utils/helpers/getParams'

import { useAppDispatch } from './useRedux'

declare type AsyncThunkConfig = {
  dispatch?: Dispatch
  extra?: unknown
  fulfilledMeta?: unknown
  pendingMeta?: unknown
  rejectValue?: unknown
  rejectedMeta?: unknown
  serializedErrorType?: unknown
  state?: unknown
}

export type UseUpdatePageProps = {
  action: AsyncThunkAction<unknown, unknown, AsyncThunkConfig>
  items: undefined | unknown[]
  page: string
  setSearchParams: SetURLSearchParams
}

const useUpdatePage = ({
  action,
  items: itemsProp,
  page: pageProp,
  setSearchParams: setSearchParamsProp,
}: UseUpdatePageProps) => {
  const dispatch = useAppDispatch()

  const updatePage = () => {
    const items = itemsProp || []
    const page = Number(pageProp)

    if (items.length === 1 && page > 1) {
      setSearchParamsProp(getParams({ page: page - 1 }))
    } else {
      dispatch(action)
    }
  }

  return { updatePage }
}

export default useUpdatePage
