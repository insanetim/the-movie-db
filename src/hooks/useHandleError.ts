import { NOTIFICATION_TYPE } from 'src/constants/app'
import { showNotification } from 'src/store/features/app'
import { useAppDispatch } from 'src/store/hooks'
import errorMessage from 'src/utils/helpers/errorMessage'

const useHandleError = () => {
  const dispatch = useAppDispatch()

  const handleError = (error: unknown) => {
    dispatch(
      showNotification({
        message: errorMessage(error),
        type: NOTIFICATION_TYPE.ERROR,
      })
    )
  }

  return { handleError }
}

export default useHandleError
