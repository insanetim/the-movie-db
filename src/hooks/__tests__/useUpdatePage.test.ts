import { act } from '@testing-library/react'
import * as reactRedux from 'src/store/hooks'
import { renderHookWithWrapper } from 'src/utils/testHelpers/renderWithWrapper'

import useUpdatePage, { UseUpdatePageProps } from '../useUpdatePage'

describe('useUpdatePage', () => {
  const mockDispatch = jest.fn()
  jest.spyOn(reactRedux, 'useAppDispatch').mockReturnValue(mockDispatch)
  const action = jest.fn()
  const setSearchParams = jest.fn()

  const props: UseUpdatePageProps = {
    action,
    items: [1],
    page: '3',
    setSearchParams,
  }

  it('should match snapshot', () => {
    const { result } = renderHookWithWrapper(() => useUpdatePage(props))

    expect(result.current).toMatchSnapshot()
  })

  it('should check "updatePage" method', () => {
    const { result } = renderHookWithWrapper(() => useUpdatePage(props))

    act(() => {
      result.current.updatePage()
    })

    expect(setSearchParams).toHaveBeenCalledWith({ page: '2' })
  })

  it('should check "updatePage" method without items', () => {
    const extendedProps = { ...props, items: undefined }

    const { result } = renderHookWithWrapper(() => useUpdatePage(extendedProps))

    act(() => {
      result.current.updatePage()
    })

    expect(mockDispatch).toHaveBeenCalledWith(action)
  })
})
