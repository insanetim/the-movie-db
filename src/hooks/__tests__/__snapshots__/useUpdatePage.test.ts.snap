// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useUpdatePage should match snapshot 1`] = `
{
  "updatePage": [Function],
}
`;
